### PR TITLE
tracker blocking reference

### PR DESCRIPTION
**Pull request by jdorweiler**
Tracker blocking reference implementation. 

To test:
`npm test` and see `test/trackers.js`

### CODE REVIEW

File: grade-scripts/shared/list-manager.js [hunk 1]
```diff
@@ -44,7 +44,8 @@ let getSymlinkedLocalList = (fileName) => {
     try {
         list = fs.readFileSync(path, { encoding: 'utf8' })
     } catch (e) {
-        throw new Error(`couldn't find and parse list ${fileName}, tried looking in: ${path}`)
+        console.warn(`couldn't find and parse list ${fileName}, tried looking in: ${path}`)
```
Comment (jdorweiler) [inline grade-scripts/shared/list-manager.js]: We don't need these files for testing the new algo so I changed it to warn instead of error.

File: grade-scripts/trackers-from-requests.js [hunk 2]
```diff
@@ -51,23 +49,44 @@ const run = async () => {
         let trackersBlocked = {}
         let trackersNotBlocked = {}
         let totalBlocked = 0
+        let requestsBlocked = []
+        let rulesUsed = {}
```
Comment (jdorweiler) [inline grade-scripts/trackers-from-requests.js]: All the changes in this file are just to collect data on what we're blocking or not blocking for use later on.

File: src/trackers.js [hunk 3]
```diff
@@ -1,227 +1,254 @@
...
+            let pattern = firstLine.split(' ')[0]
+            let b64surrogate = btoa(lines.join('\n'))
+            surrogateList[pattern] = b64dataheader + b64surrogate
+        })
+        return surrogateList
+    }
 
-        let urlSplit = hostnameToCheck.split('.')
+    getTrackerData (urlToCheck, siteUrl, request, ops) {
+        ops = ops || {}
```
Comment (jdorweiler) [inline src/trackers.js]: Not used but this is where we would pass in settings in the real implementation.

File: src/trackers.js [hunk 4]
```diff
@@ -1,227 +1,255 @@
...
-        let result = false
-        let match
+    /*
+     * Pull subdomains off of the reqeust rule and look for a matching tracker object in our data
+     */
+    findTracker (tracker, requestData) {
+        let urlList = Object.assign([], requestData.urlToCheckSplit)
 
-        match = this.checkABPParsedList(this.whitelist, url, currLocationDomain, requestType)
+        while (urlList.length > 1) {
```
Comment (kdzwinel) [inline src/trackers.js]: this doesn't take eTLDs into account, so we will check things like `co.uk`, but I guess it's not a big deal?

File: src/trackers.js [hunk 5]
```diff
@@ -1,227 +1,255 @@
...
     }
 
-    checkWhitelist (url, currLocationDomain, requestType) {
-        let result = false
-        let match
+    /*
+     * Pull subdomains off of the reqeust rule and look for a matching tracker object in our data
+     */
+    findTracker (tracker, requestData) {
+        let urlList = Object.assign([], requestData.urlToCheckSplit)
```
Comment (kdzwinel) [inline src/trackers.js]: We can use `Array.from(requestData.urlToCheckSplit)` here which is a bit more modern version of the same thing

File: src/trackers.js [hunk 6]
```diff
@@ -1,227 +1,255 @@
...
+    */
+    matchesException (tracker, requestData) {
+        if (tracker.matchedRule && tracker.matchedRule.exceptions) {
+            // all exception types need to match for this to return true
+            return !this.exceptionTypes.some(exceptionType => {
+                const requestParamToCheck = exceptionType === 'domains' ? requestData.siteDomain : requestData.request.type
+                return !(tracker.matchedRule.exceptions[exceptionType] &&
+                          tracker.matchedRule.exceptions[exceptionType].length &&
+                          tracker.matchedRule.exceptions[exceptionType].includes(requestParamToCheck))
+            })
```
Comment (kdzwinel) [inline src/trackers.js]: this is pretty much the same as `matchRuleOptions`, can we reuse `matchRuleOptions`?

File: src/trackers.js [hunk 7]
```diff
@@ -1,227 +1,255 @@
...
     }
 
-    checkABPParsedList (list, url, currLocationDomain, requestType) {
-        let match = abp.matches(list, url,
-            {
-                domain: currLocationDomain,
-                elementTypeMask: abp.elementTypes[requestType.toUpperCase()]
-            })
-        return match
+    setAction (tracker, requestData) {
```
Comment (kdzwinel) [inline src/trackers.js]: this method is pretty awesome 🙌I like how we now have one center piece where everything comes togheter

Comment (kdzwinel) [reply src/trackers.js]: nit: `requestData` is not used here

File: src/trackers.js [hunk 8]
```diff
@@ -1,227 +1,255 @@
...
+        return tracker
     }
 
-    checkWhitelist (url, currLocationDomain, requestType) {
-        let result = false
-        let match
+    /*
+     * Pull subdomains off of the reqeust rule and look for a matching tracker object in our data
+     */
+    findTracker (tracker, requestData) {
```
Comment (kdzwinel) [inline src/trackers.js]: `tracker` is not used here

File: src/trackers.js [hunk 9]
```diff
@@ -1,227 +1,255 @@
...
-        let trackerByParentCompany = this.checkTrackersWithParentCompany(urlSplit, currLocation)
-        if (trackerByParentCompany) {
-            return trackerByParentCompany
-        }
+        // compare the site owner to the tracker owner and set firstParty
+        tracker.firstParty = this.setFirstParty(tracker, requestData)
+
+        this.setAction(tracker, requestData)
+
+        return tracker
```
Comment (kdzwinel) [inline src/trackers.js]: after looking at this a bit longer I do find it a bit hard to follow that some of those methods are modifying `tracker` (often in unexpected ways) and some return a value that gets assigned to tracker property. I'd change that to be sth like this:

```js
const tracker = this.findTracker(request);

if(!tracker) {
  return null;
}

const matchedRule = this.findMatchingRule(tracker.rules, request);

if (matchedRule && matchedRule.surrogate) {
   // getSurrogateURL could find given surrogate and add the 'data:' prefix
   const redirectUrl = this.getSurrogateURL(matchedRule.surrogate);
}

// we should be able to reuse `findRule` for exceptions
const matchedRuleException = this.findMatchingRule(tracker.exceptions, request);

const trackerOwner = this.findTrackerOwner(tracker);

const websiteOwner = this.findWebsiteOwner(request);

const firstParty = trackerOwner === websiteOwner;

// only pass info that it needs
const {action, reason} = this.shouldBlock({
  firstParty,
  matchedRule,
  matchedRuleException,
  defaultAction: tracker.defaultAction,
  redirectUrl
});

// only return info that we need 
return {
  action,
  reason,
  redirectUrl,
  …
};
```

File: src/trackers.js (hunk 4 above)
Comment (jdorweiler) [reply src/trackers.js]: Yeah no big deal.

File: src/trackers.js (hunk 6 above)
Comment (jdorweiler) [reply src/trackers.js]: Added a more general function `matchesRuleDefinition` that handles both options and exceptions.

File: grade-scripts/trackers-from-requests.js [hunk 10]
```diff
@@ -51,23 +49,44 @@ const run = async () => {
...
-            let tracker = trackers.isTracker(request[0], url, request[1])
+            const t = process.hrtime()
+            const tracker = trackers.getTrackerData(request[0], url, {url: request[0], type: request[1]})
+            const time = process.hrtime(t);
 
             if (tracker) {
-                if (tracker.block) {
-                    if (!trackersBlocked[tracker.parentCompany]) {
-                        trackersBlocked[tracker.parentCompany] = {}
+                tracker.time = time
```
[the remark this answers is not in the document]
Comment (jdorweiler) [reply grade-scripts/trackers-from-requests.js]: I moved out. I'm keeping track of overall time to run the full data set now.

File: src/trackers.js [hunk 11]
```diff
@@ -1,227 +1,255 @@
...
+        this.surrogateList = this.processSurrogateList(lists.surrogates)
+        this.exceptionTypes = ['domains', 'types']
+    }
 
-        abp.parse(lists.whitelist, this.whitelist)
+    processTrackerList (data) {
+        Object.keys(data).forEach(name => {
+            if (data[name].rules) {
+                for (let i in data[name].rules) {
+                    // fix this later. make a toString to serialize RegExp?
```
[the remark this answers is not in the document]
Comment (jdorweiler) [reply src/trackers.js]: Nice, I got rid of it. I was thinking I had to write my own toString :)

File: src/trackers.js [hunk 12]
```diff
@@ -1,227 +1,255 @@
...
+            // remove first line, store it
+            let firstLine = lines.shift()
 
-        if (!hostnameToCheck) {
-            return false
-        }
+            // take identifier from first line
+            let pattern = firstLine.split(' ')[0]
+            let b64surrogate = btoa(lines.join('\n'))
+            surrogateList[pattern] = b64dataheader + b64surrogate
```
[the remark this answers is not in the document]
Comment (jdorweiler) [reply src/trackers.js]: I'll leave it like this for now.
